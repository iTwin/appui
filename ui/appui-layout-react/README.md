# @itwin/appui-layout-react

Copyright Â© Bentley Systems, Incorporated. All rights reserved. See LICENSE.md for license terms and full copyright notice.

## Description

The __@itwin/appui-layout-react__ package contains internal React components for the AppUi-react panels, widget UI and other purposes and is not expected to be used directly by applications.

## Documentation

See the [iTwin.js](https://www.itwinjs.org/learning/ui/) documentation for more information.

### Coding Guidelines

Look at the existing code and try to keep your code similar.

* Every component class name should follow nz-path-to-component-componentName naming style.

* Every inner class name should have nz- prefix.

* Import the Sass classnames with: `import "./YourComponent.scss";`

* Every component should accept `className?: string` and `style?: React.CSSProperties` props.

* Components that accept one or a list of children should use the `children?: React.ReactNode` prop.

### Advice

* Use [classnames](https://www.npmjs.com/package/classnames) function to create className strings for the elements.

* Use `const className` inside render for the root element className value.
