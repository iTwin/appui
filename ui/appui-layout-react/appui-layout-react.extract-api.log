Invoking: betools extract-api --entry=appui-layout-react 
Running command:
node "D:\appui\common\temp\node_modules\.pnpm\@itwin+build-tools@4.0.0-dev.87_@types+node@18.11.5\node_modules\@itwin\build-tools\scripts\extract-api.js" --entry appui-layout-react

api-extractor 7.34.4  - https://api-extractor.com/

Analysis will use the bundled TypeScript version 5.0.2

API Extractor completed successfully
