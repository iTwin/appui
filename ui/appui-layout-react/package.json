{
  "name": "@itwin/appui-layout-react",
  "version": "4.2.0-dev.0",
  "description": "iTwin.js Nine-zone React UI components",
  "main": "lib/cjs/appui-layout-react.js",
  "module": "lib/esm/appui-layout-react.js",
  "typings": "lib/cjs/appui-layout-react",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/iTwin/appui.git",
    "directory": "ui/appui-layout-react"
  },
  "scripts": {
    "start": "run-p -l \"build:** -- -w\" \"copy:** -- -w\"",
    "build": "npm run -s build:cjs && npm run -s build:esm",
    "copy:cjs": "cpx \"./src/**/*.{*css,svg}\" \"./lib/cjs\"",
    "copy:esm": "cpx \"./src/**/*.{*css,svg}\" \"./lib/esm\"",
    "build:cjs": "npm run -s copy:cjs && tsc 1>&2 --outDir lib/cjs",
    "build:esm": "npm run -s copy:esm && tsc 1>&2 --module ES2020 --outDir lib/esm",
    "clean": "rimraf lib .rush/temp/package-deps*.json",
    "cover": "nyc npm -s test",
    "extract-api": "betools extract-api --entry=appui-layout-react",
    "lint": "eslint -f visualstudio \"./src/**/*.{ts,tsx}\" 1>&2",
    "lint:fix": "npm run -s lint -- --fix",
    "test": "mocha --config ../.mocharc.json \"./lib/cjs/test/**/*.test.js\"",
    "test:watch": "npm -s test -- --reporter min --watch-extensions ts,tsx --watch",
    "docs": "npm run -s docs:reference && npm run -s docs:changelog",
    "docs:changelog": "cpx ./CHANGELOG.md ../../generated-docs/reference/appui-layout-react",
    "docs:reference": "betools docs --includes=../../generated-docs/extract --json=../../generated-docs/reference/appui-layout-react/file.json --excludeGlob=**/declarations.d.ts --tsIndexFile=./appui-layout-react.ts --onlyJson"
  },
  "keywords": [
    "AppUI",
    "Bentley",
    "BIM",
    "iModel"
  ],
  "author": {
    "name": "Bentley Systems, Inc.",
    "url": "http://www.bentley.com"
  },
  "peerDependencies": {
    "@itwin/appui-abstract": "^3.7.0 || ^4.0.0",
    "@itwin/core-bentley": "^3.7.0 || ^4.0.0",
    "@itwin/core-react": "workspace:^4.2.0-dev.0",
    "react": "^17.0.0 || ^18.0.0",
    "react-dom": "^17.0.0 || ^18.0.0"
  },
  "//devDependencies": [
    "NOTE: All peerDependencies should also be listed as devDependencies since peerDependencies are not considered by npm install",
    "NOTE: All tools used by scripts in this package must be listed as devDependencies"
  ],
  "devDependencies": {
    "@itwin/appui-abstract": "^3.7.0",
    "@itwin/build-tools": "^3.7.0",
    "@itwin/core-bentley": "^3.7.0",
    "@itwin/core-react": "workspace:*",
    "@itwin/eslint-plugin": "^3.7.0",
    "@testing-library/dom": "^8.11.2",
    "@testing-library/react": "^12.0.0",
    "@testing-library/react-hooks": "^7.0.2",
    "@testing-library/user-event": "^14.4.2",
    "@types/chai": "4.3.1",
    "@types/chai-as-promised": "^7",
    "@types/chai-jest-snapshot": "^1.3.0",
    "@types/lodash": "^4.14.0",
    "@types/mocha": "^8.2.2",
    "@types/node": "18.11.5",
    "@types/react": "^17.0.37",
    "@types/react-dom": "^17.0.0",
    "@types/react-transition-group": "^4.4.4",
    "@types/sinon": "^9.0.0",
    "chai": "^4.1.2",
    "chai-as-promised": "^7",
    "chai-jest-snapshot": "^2.0.0",
    "cpx2": "^3.0.0",
    "eslint": "^7.11.0",
    "ignore-styles": "^5.0.1",
    "jsdom": "^19.0.0",
    "jsdom-global": "3.0.2",
    "mocha": "^10.0.0",
    "npm-run-all": "^4.1.5",
    "nyc": "^15.1.0",
    "raf": "^3.4.0",
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "rimraf": "^3.0.2",
    "sinon": "^9.0.2",
    "sinon-chai": "^3.2.0",
    "ts-node": "^10.8.2",
    "typemoq": "^2.1.0",
    "typescript": "~5.0.2",
    "eslint-config-prettier": "~8.8.0",
    "@rushstack/eslint-patch": "^1.3.0"
  },
  "//dependencies": [
    "NOTE: these dependencies should be only for things that DO NOT APPEAR IN THE API",
    "NOTE: core-frontend should remain UI technology agnostic, so no react/angular dependencies are allowed"
  ],
  "dependencies": {
    "@itwin/itwinui-variables": "^2.0.0",
    "@itwin/itwinui-icons-react": "^2.1.0",
    "classnames": "2.3.1",
    "immer": "9.0.6",
    "lodash": "^4.17.10",
    "react-transition-group": "^4.4.2",
    "zustand": "^4.3.6"
  },
  "nyc": {
    "extends": "./node_modules/@itwin/build-tools/.nycrc",
    "require": [
      "ignore-styles",
      "jsdom-global/register",
      "source-map-support/register",
      "ts-node/register"
    ],
    "check-coverage": true,
    "statements": 100,
    "branches": 100,
    "functions": 100,
    "lines": 100
  }
}