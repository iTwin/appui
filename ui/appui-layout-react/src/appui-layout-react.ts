/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
export * from "./appui-layout-react/backstage/Backstage";
export * from "./appui-layout-react/backstage/Item";
export * from "./appui-layout-react/backstage/Separator";

export * from "./appui-layout-react/base/DragManager";
export * from "./appui-layout-react/base/Ellipsis";
export * from "./appui-layout-react/base/NineZone";
export * from "./appui-layout-react/base/SafeAreaInsets";
export * from "./appui-layout-react/base/usePointerCaptor";

export * from "./appui-layout-react/footer/dialog/Button";
export * from "./appui-layout-react/footer/dialog/Dialog";
export * from "./appui-layout-react/footer/dialog/TitleBar";

export * from "./appui-layout-react/footer/message/Button";
export * from "./appui-layout-react/footer/message/Hyperlink";
export * from "./appui-layout-react/footer/message/Layout";
export * from "./appui-layout-react/footer/message/Message";
export * from "./appui-layout-react/footer/message/Progress";
export * from "./appui-layout-react/footer/message/Status";
export * from "./appui-layout-react/footer/message/Toast";

export * from "./appui-layout-react/footer/message-center/Dialog";
export * from "./appui-layout-react/footer/message-center/Indicator";
export * from "./appui-layout-react/footer/message-center/Message";
export * from "./appui-layout-react/footer/message-center/Tab";

export * from "./appui-layout-react/footer/snap-mode/Indicator";
export * from "./appui-layout-react/footer/snap-mode/Panel";
export * from "./appui-layout-react/footer/snap-mode/Snap";

export * from "./appui-layout-react/footer/tool-assistance/Dialog";
export * from "./appui-layout-react/footer/tool-assistance/Indicator";
export * from "./appui-layout-react/footer/tool-assistance/Instruction";
export * from "./appui-layout-react/footer/tool-assistance/Item";
export * from "./appui-layout-react/footer/tool-assistance/Separator";

export * from "./appui-layout-react/footer/Footer";
export * from "./appui-layout-react/footer/Indicator";
export * from "./appui-layout-react/footer/Popup";

export * from "./appui-layout-react/popup/Tooltip";

export * from "./appui-layout-react/state/DropTargetState";
export * from "./appui-layout-react/state/NineZoneAction";
export * from "./appui-layout-react/state/NineZoneState";
export * from "./appui-layout-react/state/NineZoneStateReducer";
export * from "./appui-layout-react/state/PanelState";
export * from "./appui-layout-react/state/TabLocation";
export * from "./appui-layout-react/state/TabState";
export * from "./appui-layout-react/state/ToolSettingsState";
export * from "./appui-layout-react/state/WidgetLocation";
export * from "./appui-layout-react/state/WidgetState";

export * from "./appui-layout-react/tool-settings/Docked";
export * from "./appui-layout-react/tool-settings/Handle";
export * from "./appui-layout-react/tool-settings/Overflow";
export * from "./appui-layout-react/tool-settings/Panel";
export * from "./appui-layout-react/tool-settings/Setting";

export * from "./appui-layout-react/widget/Buttons";
export * from "./appui-layout-react/widget/Content";
export * from "./appui-layout-react/widget/ContentContainer";
export * from "./appui-layout-react/widget/ContentManager";
export * from "./appui-layout-react/widget/ContentRenderer";
export * from "./appui-layout-react/widget/Dock";
export * from "./appui-layout-react/widget/FloatingTab";
export * from "./appui-layout-react/widget/FloatingWidget";
export * from "./appui-layout-react/widget/FloatingWidgets";
export * from "./appui-layout-react/widget/Menu";
export * from "./appui-layout-react/widget/NavigationArea";
export * from "./appui-layout-react/widget/Overflow";
export * from "./appui-layout-react/widget/PanelWidget";
export * from "./appui-layout-react/widget/PinToggle";
export * from "./appui-layout-react/widget/PopoutToggle";
export * from "./appui-layout-react/widget/SendBack";
export * from "./appui-layout-react/widget/Tab";
export * from "./appui-layout-react/widget/TabBar";
export * from "./appui-layout-react/widget/Tabs";
export * from "./appui-layout-react/widget/ToolsArea";
export * from "./appui-layout-react/widget/Widget";

export * from "./appui-layout-react/widget/tools/button/App";
export * from "./appui-layout-react/widget/tools/button/Back";
export * from "./appui-layout-react/widget/tools/button/Button";
export * from "./appui-layout-react/widget/tools/button/Expandable";
export * from "./appui-layout-react/widget/tools/button/Icon";

export * from "./appui-layout-react/widget-panels/AppContent";
export * from "./appui-layout-react/widget-panels/CenterContent";
export * from "./appui-layout-react/widget-panels/Content";
export * from "./appui-layout-react/widget-panels/CursorOverlay";
export * from "./appui-layout-react/widget-panels/Expander";
export * from "./appui-layout-react/widget-panels/Grip";
export * from "./appui-layout-react/widget-panels/Panel";
export * from "./appui-layout-react/widget-panels/Panels";

/** @docs-package-description
 * The appui-layout-react package contains React components for application user interface layouts following the Bentley 9-Zone pattern.
 */
/**
 * @docs-group-description Backstage
 * Classes for working with a Backstage
 */
/**
 * @docs-group-description Footer
 * Classes for working with a Footer or Status Bar
 */
/**
 * @docs-group-description Message
 * Classes for working with Messages
 */
/**
 * @docs-group-description MessageCenter
 * Classes for working with the MessageCenter
 */
/**
 * @docs-group-description Popup
 * Classes for working with a Popup
 */
/**
 * @docs-group-description SnapMode
 * Classes for working with SnapMode UI
 */
/**
 * @docs-group-description ToolAssistance
 * Classes for working with Tool Assistance
 */
/**
 * @docs-group-description Toolbar
 * Classes for working with a Toolbar
 */
/**
 * @docs-group-description ToolSettings
 * Classes for working with Tool Settings
 */
/**
 * @docs-group-description Utilities
 * Utility classes for working with the 9-Zone pattern
 */
/**
 * @docs-group-description Widget
 * Classes for working a Widget
 */
/**
 * @docs-group-description WidgetPanels
 * Classes for working with widget panels
 */
