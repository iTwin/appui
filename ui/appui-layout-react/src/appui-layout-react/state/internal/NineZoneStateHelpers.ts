/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import type { Draft } from "immer";
import type { PointProps } from "@itwin/appui-abstract";
import type { NineZoneState } from "../NineZoneState";
import type {
  FloatingWidgetHomeState,
  FloatingWidgetState,
} from "../WidgetState";
import type { RectangleProps, SizeProps } from "@itwin/core-react";
import { Rectangle } from "@itwin/core-react";
import {
  getWidgetState,
  updateFloatingWidgetState,
} from "./WidgetStateHelpers";

/** @internal */
export const category = "appui-layout-react:layout";

/** Helper that returns a plain object. Prevents from adding an instance that is not `immerable` to the state.
 * @internal
 */
export function toRectangleProps(
  rectangle: RectangleProps | undefined
): RectangleProps {
  if (rectangle) {
    return {
      bottom: rectangle.bottom,
      left: rectangle.left,
      right: rectangle.right,
      top: rectangle.top,
    };
  }
  return new Rectangle().toProps();
}

/** @internal */
export function setRectangleProps(
  props: Draft<RectangleProps>,
  bounds: RectangleProps
) {
  props.left = bounds.left;
  props.right = bounds.right;
  props.top = bounds.top;
  props.bottom = bounds.bottom;
}

/** @internal */
export function setPointProps(props: Draft<PointProps>, point: PointProps) {
  props.x = point.x;
  props.y = point.y;
}

/** @internal */
export function setSizeProps(props: Draft<SizeProps>, size: SizeProps) {
  props.height = size.height;
  props.width = size.width;
}

type KeysOfType<T, Type> = {
  [K in keyof T]: T[K] extends Type ? K : never;
}[keyof T];

/** @internal */
export function initSizeProps<
  T,
  K extends KeysOfType<T, SizeProps | undefined>
>(obj: T, key: K, size: SizeProps | undefined) {
  if (obj[key] && size) {
    setSizeProps(obj[key] as unknown as SizeProps, size);
    return;
  }
  (obj[key] as unknown as SizeProps | undefined) = size
    ? {
        height: size.height,
        width: size.width,
      }
    : undefined;
}

/** @internal */
export function initRectangleProps<
  T,
  K extends KeysOfType<T, RectangleProps | undefined>
>(obj: T, key: K, rectangle: T[K]) {
  const rect = rectangle as unknown as RectangleProps | undefined;
  if (obj[key] && rect) {
    setRectangleProps(obj[key] as unknown as RectangleProps, rect);
    return;
  }
  (obj[key] as unknown as RectangleProps | undefined) = rect
    ? {
        bottom: rect.bottom,
        left: rect.left,
        right: rect.right,
        top: rect.top,
      }
    : undefined;
}

/** @internal */
export function isToolSettingsFloatingWidget(
  state: NineZoneState,
  id: FloatingWidgetState["id"]
) {
  const widget = getWidgetState(state, id);
  const toolSettingsTabId = state.toolSettings?.tabId;
  return (
    widget.tabs.length === 1 &&
    widget.tabs[0] === toolSettingsTabId &&
    id in state.floatingWidgets.byId
  );
}

/** @internal */
export function updateHomeOfToolSettingsWidget(
  state: NineZoneState,
  id: FloatingWidgetState["id"],
  home: FloatingWidgetHomeState
): NineZoneState {
  if (!isToolSettingsFloatingWidget(state, id)) return state;

  return updateFloatingWidgetState(state, id, {
    home,
  });
}
