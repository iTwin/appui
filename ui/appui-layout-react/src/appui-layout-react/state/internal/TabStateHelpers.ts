/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import produce from "immer";
import { UiError } from "@itwin/appui-abstract";
import type { NineZoneState } from "../NineZoneState";
import type { DraggedTabState, TabState } from "../TabState";
import { category } from "./NineZoneStateHelpers";
import type { SavedTabState } from "../SavedTabState";
import type { WritableDraft } from "immer/dist/internal";

/** @internal */
export function createTabState(
  id: TabState["id"],
  args?: Partial<TabState>
): TabState {
  return {
    label: "",
    ...args,
    id,
    unloaded: false,
  };
}

/** @internal */
export function createDraggedTabState(
  tabId: DraggedTabState["tabId"],
  args?: Partial<DraggedTabState>
): DraggedTabState {
  return {
    home: {
      side: "left",
      widgetId: undefined,
      widgetIndex: 0,
    },
    position: { x: 0, y: 0 },
    ...args,
    tabId,
  };
}

/** @internal */
export function updateTabState(
  state: NineZoneState,
  id: TabState["id"],
  update: (draft: WritableDraft<TabState>) => void
) {
  if (!(id in state.tabs)) throw new UiError(category, "Tab does not exist");

  return produce(state, (draft) => {
    const tab = draft.tabs[id];
    update(tab);
  });
}

/** @internal */
export function updateSavedTabState(
  state: NineZoneState,
  id: TabState["id"],
  update: (draft: WritableDraft<SavedTabState>) => void
) {
  return produce(state, (draft) => {
    const allIds = draft.savedTabs.allIds;
    const byId = draft.savedTabs.byId;
    let tab = byId[id];
    if (!tab) {
      allIds.push(id);
      tab = byId[id] = { id };
    } else {
      const index = allIds.indexOf(id);
      allIds.splice(index, 1);
      allIds.push(id);
    }

    update(tab);
  });
}
