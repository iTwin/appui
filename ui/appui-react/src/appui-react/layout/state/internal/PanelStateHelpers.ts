/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Base
 */

import { UiError } from "@itwin/appui-abstract";
import type { Draft } from "immer";
import { produce } from "immer";
import { isHorizontalPanelSide } from "../../widget-panels/Panel.js";
import type {
  HorizontalPanelSide,
  PanelSide,
  VerticalPanelSide,
} from "../../widget-panels/PanelTypes.js";
import type { NineZoneState } from "../NineZoneState.js";
import type {
  HorizontalPanelState,
  PanelsState,
  VerticalPanelState,
} from "../PanelState.js";
import type { WidgetState } from "../WidgetState.js";
import { category } from "./NineZoneStateHelpers.js";
import { addWidgetState } from "./WidgetStateHelpers.js";
import type { StagePanelSizeSpec } from "../../../stagepanels/StagePanelConfig.js";
import type { SizeProps } from "../../../utils/SizeProps.js";

function createPanelState(side: PanelSide) {
  return {
    collapseOffset: 100,
    collapsed: false,
    maxSize: 600,
    minSize: 200,
    pinned: true,
    resizable: true,
    side,
    size: undefined,
    widgets: [],
    maxWidgetCount: 2,
    splitterPercent: 50,
  };
}

/** @internal */
export function createVerticalPanelState(
  side: VerticalPanelSide,
  args?: Partial<VerticalPanelState>
): VerticalPanelState {
  return {
    ...createPanelState(side),
    ...args,
    side,
  };
}

/** @internal */
export function createHorizontalPanelState(
  side: HorizontalPanelSide,
  args?: Partial<HorizontalPanelState>
): HorizontalPanelState {
  return {
    ...createPanelState(side),
    minSize: 100,
    span: true,
    ...args,
    side,
  };
}

/** @internal */
export function createPanelsState(args?: Partial<PanelsState>): PanelsState {
  return {
    bottom: createHorizontalPanelState("bottom"),
    left: createVerticalPanelState("left"),
    right: createVerticalPanelState("right"),
    top: createHorizontalPanelState("top"),
    ...args,
  };
}

/** @internal */
export function updatePanelState<K extends keyof PanelsState>(
  state: NineZoneState,
  side: K,
  update: (draft: Draft<PanelsState>[K]) => void
) {
  return produce(state, (draft) => {
    const panel = draft.panels[side];
    update(panel);
  });
}

/** @internal */
export function getPanelPixelSizeFromSpec(
  side: PanelSide,
  appSize: SizeProps,
  panelSize: StagePanelSizeSpec
) {
  if (typeof panelSize === "number") {
    return panelSize;
  }

  const fullSize = isHorizontalPanelSide(side) ? appSize.height : appSize.width;
  return (panelSize.percentage / 100) * fullSize;
}

/** @internal */
export function getPanelSize(
  preferredSizeSpec: StagePanelSizeSpec | undefined,
  side: PanelSide,
  minSizeSpec: StagePanelSizeSpec,
  maxSizeSpec: StagePanelSizeSpec,
  appSize: SizeProps
) {
  if (preferredSizeSpec === undefined) return undefined;

  const maxSize = getPanelPixelSizeFromSpec(side, appSize, maxSizeSpec);
  const minSize = getPanelPixelSizeFromSpec(side, appSize, minSizeSpec);
  const preferredSize = getPanelPixelSizeFromSpec(
    side,
    appSize,
    preferredSizeSpec
  );

  return Math.min(Math.max(preferredSize, minSize), maxSize);
}

/** @internal */
export function addPanelWidget(
  state: NineZoneState,
  side: PanelSide,
  id: WidgetState["id"],
  tabs: WidgetState["tabs"],
  widgetArgs?: Partial<WidgetState>
): NineZoneState {
  return insertPanelWidget(state, side, id, tabs, Infinity, widgetArgs);
}

/** @internal */
export function insertPanelWidget(
  state: NineZoneState,
  side: PanelSide,
  id: WidgetState["id"],
  tabs: WidgetState["tabs"],
  sectionIndex: number,
  widgetArgs?: Partial<WidgetState>
): NineZoneState {
  const panel = state.panels[side];
  const maxWidgetCount = panel.maxWidgetCount;
  if (panel.widgets.length >= maxWidgetCount)
    // eslint-disable-next-line @typescript-eslint/no-deprecated
    throw new UiError(category, "Max widget count exceeded", undefined, () => ({
      maxWidgetCount,
    }));

  state = addWidgetState(state, id, tabs, widgetArgs);
  return produce(state, (draft) => {
    const widgets = draft.panels[side].widgets;
    widgets.splice(sectionIndex, 0, id);
  });
}
