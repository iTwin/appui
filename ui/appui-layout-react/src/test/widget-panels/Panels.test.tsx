/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import * as React from "react";
import { render } from "@testing-library/react";
import { createNineZoneState, WidgetPanels } from "../../appui-layout-react";
import { TestNineZoneProvider } from "../Providers";
import { addPanelWidget } from "../../appui-layout-react/state/internal/PanelStateHelpers";
import { addTab } from "../../appui-layout-react/state/internal/TabStateHelpers";
import { addWidgetToolSettings } from "../../appui-layout-react/state/internal/ToolSettingsStateHelpers";

describe("WidgetPanels", () => {
  it("should render", () => {
    const { container } = render(
      <TestNineZoneProvider>
        <WidgetPanels />
      </TestNineZoneProvider>
    );
    container.firstChild!.should.matchSnapshot();
  });

  it("should render widget content", async () => {
    let state = createNineZoneState();
    state = addTab(state, "t1");
    state = addPanelWidget(state, "left", "w1", ["t1"]);
    const { findByText } = render(
      <TestNineZoneProvider
        defaultState={state}
        widgetContent={<div>Hello Widget!</div>}
      >
        <WidgetPanels />
      </TestNineZoneProvider>
    );
    await findByText("Hello Widget!");
  });

  it("should render tool settings content", async () => {
    let state = createNineZoneState();
    state = addTab(state, "ts");
    state = addPanelWidget(state, "left", "w1", ["ts"]);
    state = addWidgetToolSettings(state, "ts");
    const { findByText } = render(
      <TestNineZoneProvider
        defaultState={state}
        toolSettingsContent={<div>Hello ToolSettings!</div>}
      >
        <WidgetPanels />
      </TestNineZoneProvider>
    );
    await findByText("Hello ToolSettings!");
  });
});
