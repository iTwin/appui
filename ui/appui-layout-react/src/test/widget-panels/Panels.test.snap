// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetPanels should render 1`] = `
<div
  class="nz-widgetPanels-panels"
>
  <div
    class="nz-widgetPanels-content nz-pinned-left nz-pinned-right nz-pinned-top nz-pinned-bottom nz-widgetPanels-appContent"
  />
  <div
    class="nz-widgetPanels-content nz-pinned-left nz-pinned-right nz-pinned-top nz-pinned-bottom nz-widgetPanels-centerContent"
  />
  <div
    class="nz-target-targetContainer nz-vertical nz-target-panelTargets nz-left"
  >
    <div
      class="nz-target-panelTarget nz-vertical nz-hidden"
    />
  </div>
  <div
    class="nz-outline-panelOutline nz-hidden nz-left"
    style="width: 200px;"
  />
  <div
    class="nz-target-targetContainer nz-vertical nz-target-panelTargets nz-right"
  >
    <div
      class="nz-target-panelTarget nz-vertical nz-hidden"
    />
  </div>
  <div
    class="nz-outline-panelOutline nz-hidden nz-right"
    style="width: 200px;"
  />
  <div
    class="nz-target-targetContainer nz-horizontal nz-target-panelTargets nz-top nz-span"
  >
    <div
      class="nz-target-panelTarget nz-horizontal nz-hidden"
    />
  </div>
  <div
    class="nz-outline-panelOutline nz-hidden nz-top nz-span"
    style="height: 100px;"
  />
  <div
    class="nz-target-targetContainer nz-horizontal nz-target-panelTargets nz-bottom nz-span"
  >
    <div
      class="nz-target-panelTarget nz-horizontal nz-hidden"
    />
  </div>
  <div
    class="nz-outline-panelOutline nz-hidden nz-bottom nz-span"
    style="height: 100px;"
  />
</div>
`;
